src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -3231,8 +3231,14 @@ static EndpointsByReplica getChangedReplicasForLeaving(String keyspaceName, Inet
 
     public void onJoin(InetAddressAndPort endpoint, EndpointState epState)
     {
-        for (Map.Entry<ApplicationState, VersionedValue> entry : epState.states())
+        Set<Map.Entry<ApplicationState, VersionedValue>> states = epState.states();
+        boolean hasStatusAndPort = states.stream()
+                                         .anyMatch(entry -> entry.getKey() == ApplicationState.STATUS_WITH_PORT);
+        for (Map.Entry<ApplicationState, VersionedValue> entry : states)
         {
+            if (hasStatusAndPort && entry.getKey() == ApplicationState.STATUS) // avoid double-processing this endpoint for things like move/remove
```

Comment:
nit: I would rename this to some `statusProcessed` construct as it's a bit weird to read?

Reply:
Are you suggesting that we just process either component and call it a day? I had it in my head that I should favor the new STATUS_AND_PORT so I was flagging its availability and bypassing legacy STATUS. Not sure if that is warranted.

Reply:
I think we need to make sure it processes STATUS_AND_PORT when present to make BOOTSTRAP_AND_REPLACE work properly when there are multiple nodes per host.

Reply:
Nah I was just suggesting changing the implementation for better readability. The functionality lgtm. In any case it's just a nit.